_posts/2021-02-04-do-not-trust-the-user-mental-model.md

```diff
@@ -0,0 +1,170 @@
+
+> There are many cases in which a data model-based design works flawlessly. Don't get me wrong; I'm not implying that data model-based design approaches are a mistake. There is nothing wrong with building an application that uses something like "Active Record," whose foundations are within a data model-based design.
+
+Let's use the well-known and sometimes abused e-commerce domain. In the e-commerce domain, there is the concept of customers. Customers want to browse for Products to select what best suits their needs.
+
+A data model-based design would quickly end up with the following (pseudo-code) entities:
+
+```
+entity Customer
+{
+   string Name;
+   string Address;
+}
+
+entity Product
+{
+   string Name;
+   string Description;
+   decimal Price;
+}
+``` 
+
+The resulting model is also relational databases friendly; it's simple to translate into tables, and more importantly, it also seems to be well aligned with the user mental model.
+
+Human beings store information as "still images." It's easier to remember something stable and well known, rather than continually changing and not easy to classify. With this in mind, let's dive into an exciting yet very simple use-case: Premium customers get a 5% discount on their purchases.
+
+Given that we already have a model, the natural reaction is to fit the new requirement into the existing model:
+
+```
+entity Customer
+{
+   string Name;
+   string Address;
+   bool IsPremium;
+}
+```
+
+Adding the attribute to the Customer entity sounds like an innocent trivial change. At checkout time, the system will check if the customer is a premium one and apply the discount according to the rule.
+
+## Please welcome Service Boundaries
+
+Let's take the above-defined model and move it in the context of a distributed system. We'll probably end up saying that Customers belong to a different service than Products, for example:
+
+- The Customers Relationships owns the Customer concept
+- the Sales service owns the Product concept.
+
+> Don't hang up too much on names. They are not the point of the article.
+
+Assuming the above is correct, we immediately have a problem. To satisfy the new requirement, a hypothetical page showing products to our customers needs to query the customers owning service to gather the "IsPremium" attribute to determine the discount to apply to the displayed products' prices. The same applies if we move the discount calculation in the shopping cart or during the checkout phase.
+
+Regardless of where the discount calculation happens, coupling is served.
+
+## Reboot
+
+First, let's remove the names we assigned to the entities and start using fantasy names, such as colors. And instead of calling them entities, let's use the word service since the intention is to identify a set of attributes that makes sense together in the context of an owner, the service.
```

Comment:
```suggestion
First, let's remove the names we assigned to the entities and start using fantasy names, such as colors. This will help us remove any preconceived notion of the data model. And instead of calling them entities, let's use the word service since the intention is to identify a set of attributes that makes sense together in the context of an owner, the service.
```